File: app/src/main/java/pl/karol202/paintplus/tool/shape/polygon/PolygonProperties.java
package pl.karol202.paintplus.tool.shape.polygon;

import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.*;
import android.widget.CompoundButton.OnCheckedChangeListener;
import android.widget.SeekBar.OnSeekBarChangeListener;
import pl.karol202.paintplus.R;
import pl.karol202.paintplus.tool.shape.ShapeProperties;

public class PolygonProperties extends ShapeProperties
							   implements OnClickListener, TextWatcher, OnCheckedChangeListener, OnSeekBarChangeListener
{
	private final int MIN_SIDES = 3;
	private final int MAX_SIDES = 20;
	
	private ShapePolygon polygon;
	private String errorToFew;
	private String errorToMany;
	
	private View view;
	private ImageButton buttonMinusSides;
	private ImageButton buttonPlusSides;
	private TextInputLayout editLayoutSides;
	private EditText editSides;
	private CheckBox checkFill;
	private SeekBar seekBarWidth;
	private TextView textWidth;
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
	{
		super.onCreateView(inflater, container, savedInstanceState);
		view = inflater.inflate(R.layout.properties_polygon, container, false);
		
		polygon = (ShapePolygon) shape;
		errorToFew = getActivity().getString(R.string.error_polygon_too_few_sides);
		errorToMany = getActivity().getString(R.string.error_polygon_too_many_sides);
		
		buttonMinusSides = (ImageButton) view.findViewById(R.id.button_minus_polygon_sides);
		buttonMinusSides.setOnClickListener(this);
		
		buttonPlusSides = (ImageButton) view.findViewById(R.id.button_plus_polygon_sides);
		buttonPlusSides.setOnClickListener(this);
		
		editLayoutSides = (TextInputLayout) view.findViewById(R.id.edit_layout_polygon_sides);
		editLayoutSides.setHintEnabled(false);
		
		editSides = editLayoutSides.getEditText();
		if(editSides == null) throw new RuntimeException("TextInputLayout must contain EditText.");
		editSides.setText(String.valueOf(polygon.getSides()));
		editSides.addTextChangedListener(this);
		
		checkFill = (CheckBox) view.findViewById(R.id.check_polygon_fill);
		checkFill.setChecked(polygon.isFill());
		checkFill.setOnCheckedChangeListener(this);
		
		seekBarWidth = (SeekBar) view.findViewById(R.id.seek_polygon_width);
		seekBarWidth.setProgress(polygon.getLineWidth() - 1);
		seekBarWidth.setOnSeekBarChangeListener(this);
		
		textWidth = (TextView) view.findViewById(R.id.polygon_width);
		textWidth.setText(String.valueOf(polygon.getLineWidth()));
		
		return view;
	}
	
	@Override
	public void onClick(View v)
	{
		if(v == buttonMinusSides)
		{
			if(getSides() <= MIN_SIDES) return;
			editSides.setText(String.valueOf(getSides() - 1));
		}
		else if(v == buttonPlusSides)
		{
			if(getSides() >= MAX_SIDES) return;
			editSides.setText(String.valueOf(getSides() + 1));
		}
	}
	
	private int getSides()
	{
		if(editSides.getText().length() == 0) return 0;
		return Integer.parseInt(editSides.getText().toString());
	}
	
	@Override
	public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
	
	@Override
	public void onTextChanged(CharSequence s, int start, int before, int count) { }
	
	@Override
	public void afterTextChanged(Editable s)
	{
		int sides = getSides();
		if(sides < MIN_SIDES) editLayoutSides.setError(errorToFew);
		else if(sides > MAX_SIDES) editLayoutSides.setError(errorToMany);
		else
		{
			editLayoutSides.setErrorEnabled(false);
			polygon.setSides(getSides());
		}
	}
	
	@Override
	public void onCheckedChanged(CompoundButton buttonView, boolean isChecked)
	{
		polygon.setFill(isChecked);
	}
	
	@Override
	public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser)
	{
		polygon.setLineWidth(progress + 1);
		textWidth.setText(String.valueOf(progress + 1));
	}
	
	@Override
	public void onStartTrackingTouch(SeekBar seekBar) { }
	
	@Override
	public void onStopTrackingTouch(SeekBar seekBar) { }
}

File: app/src/main/java/pl/karol202/paintplus/tool/shape/polygon/PolygonProperties.kt
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package pl.karol202.paintplus.tool.shape.polygon

import android.annotation.SuppressLint
import pl.karol202.paintplus.tool.shape.JoinAdapter
import android.os.Bundle
import pl.karol202.paintplus.R
import android.view.View
import pl.karol202.paintplus.util.SeekBarTouchListener
import androidx.core.widget.addTextChangedListener
import androidx.fragment.app.Fragment
import org.koin.android.ext.android.inject
import pl.karol202.paintplus.databinding.PropertiesPolygonBinding
import pl.karol202.paintplus.tool.shape.Join
import pl.karol202.paintplus.util.setOnItemSelectedListener
import pl.karol202.paintplus.util.setOnValueChangeListener
import pl.karol202.paintplus.util.viewBinding

private const val MIN_SIDES = 3
private const val MAX_SIDES = 20

class PolygonProperties : Fragment(R.layout.properties_polygon)
{
	private val shapePolygon by inject<ShapePolygon>()

	private val views by viewBinding(PropertiesPolygonBinding::bind)

	private var errorToFew = getString(R.string.error_polygon_too_few_sides)
	private var errorToMany = getString(R.string.error_polygon_too_many_sides)

	private val sides get() = views.editTextPolygonSides.text?.toString()?.toIntOrNull() ?: 0

	@SuppressLint("ClickableViewAccessibility")
	override fun onViewCreated(view: View, savedInstanceState: Bundle?)
	{
		views.buttonMinusPolygonSides.setOnClickListener { onSidesDecremented() }

		views.buttonPlusPolygonSides.setOnClickListener { onSidesIncremented() }

		views.editLayoutPolygonSides.isHintEnabled = false
		views.editTextPolygonSides.setText(shapePolygon.sides.toString())
		views.editTextPolygonSides.addTextChangedListener { onSidesChanged() }

		views.checkPolygonFill.isChecked = shapePolygon.isFilled
		views.checkPolygonFill.setOnCheckedChangeListener { _, checked: Boolean -> onFillChanged(checked) }

		views.seekPolygonWidth.progress = shapePolygon.lineWidth - 1
		views.seekPolygonWidth.setOnValueChangeListener { onWidthChanged(it) }
		views.seekPolygonWidth.setOnTouchListener(SeekBarTouchListener())

		views.polygonWidth.text = shapePolygon.lineWidth.toString()

		views.spinnerPolygonJoin.adapter = JoinAdapter(requireContext())
		views.spinnerPolygonJoin.setSelection(shapePolygon.join.ordinal)
		views.spinnerPolygonJoin.setOnItemSelectedListener { onJoinChanged(it) }
	}

	private fun onSidesDecremented()
	{
		if(sides > MIN_SIDES) views.editTextPolygonSides.setText((sides - 1).toString())
	}

	private fun onSidesIncremented()
	{
		if(sides < MAX_SIDES) views.editTextPolygonSides.setText((sides + 1).toString())
	}

	private fun onSidesChanged() = when
	{
		sides < MIN_SIDES -> views.editLayoutPolygonSides.error = errorToFew
		sides > MAX_SIDES -> views.editLayoutPolygonSides.error = errorToMany
		else ->
		{
			views.editLayoutPolygonSides.isErrorEnabled = false
			shapePolygon.sides = sides
		}
	}

	private fun onFillChanged(fill: Boolean)
	{
		shapePolygon.isFilled = fill
	}

	private fun onWidthChanged(progress: Int)
	{
		shapePolygon.lineWidth = progress + 1
		views.polygonWidth.text = (progress + 1).toString()
	}

	private fun onJoinChanged(index: Int)
	{
		shapePolygon.join = Join.values()[index]
	}
}
